benchmark/simulate_hgts.py

```diff
@@ -0,0 +1,627 @@
+
+    Returns
+    -------
+    species_ids: dictionary
+    sequence_ids: dictionary
+    orthologous_groups: list of lists
+    """
+    species_ids = {}
+    sequence_ids = {}
+    orthologous_groups = []
+    with open(
+        glob.glob(
+            join(
+                results_dir,
+                "clusters_OrthoFinder_*_id_pairs.txt"))[0], 'r') as id_pairs_f:
+        # skip header lines
+        for _ in range(7):
+            next(id_pairs_f)
+        # parse orthologous family groups
+        for line in id_pairs_f:
+            line = line.strip().split()
+            # include only families with at least 2 orthologs
+            if len(line[1:-1]) > 1:
+                orthologous_groups.append(line[1:-1])
+    with open(join(results_dir, "SpeciesIDs.txt"), 'r') as species_ids_f:
+        for line in species_ids_f:
+            line = line.strip().split()
+            species_ids[line[0].split(':')[0]] = line[1]
+    with open(join(results_dir, "SequenceIDs.txt"), 'r') as sequence_ids_f:
+        for line in sequence_ids_f:
+            line = line.strip().split()
+            sequence_ids[line[0].split(':')[0]] = line[1]
+
+    return species_ids, sequence_ids, orthologous_groups
+
+
+def simulate_orthologous_rep(genes_donor,
+                             seq_donor,
+                             genes_recip,
+                             seq_recip,
+                             sequence_ids,
+                             orthologous_groups,
+                             orthologous_rep_prob,
+                             percentage_hgts,
+                             log_f):
+    """Simulate orthologous replacement HGT.
+
+    Notes
+    -----
+    Algorithm: Using list of orthologous genes between donor and recipient
+    genomes
+
+    Parameters
+    ----------
+    genes_donor: dictionary
+    seq_donor: skbio.sequence.Sequence
+    genes_recip: dictionary
+    seq_recip: skbio.sequence.Sequence
+    sequence_ids: dictionary
+    orthologous_groups: list of lists
+    orthologous_rep_prob: float
+    percentage_hgts: float
+    log_f: file descriptor
+
+    Returns
+    -------
+    seq_recip: skbio.sequence.Sequence
+        recipient genome sequence with HGTs
+    """
+    # number of HGTs to simulate
+    num_hgts = int(percentage_hgts*orthologous_rep_prob*len(genes_recip))
+    if num_hgts < 1:
+        num_hgts = 1
+    num_orthogroups = len(orthologous_groups)
+    idx = random.sample(list(range(0, num_orthogroups)), num_hgts)
+    log_f.write("#type\tdonor\tstart\tend\trecipient\tnew label "
+                "recipient\tstart\tend\tstrand\n")
+    for x in range(0, num_hgts):
+        orthogroup = orthologous_groups[idx[x]]
+        substitute_genes = ['*', '*']
+        # randomly select two orthologous genes from the same family
+        # representing the donor and recipient genomes
+        while '*' in substitute_genes:
+            idx2 = random.randrange(0, len(orthogroup))
```

Comment:
I added a inline note on this, however:

Each orthogroup is guranteed to have at least two genes, one from donor (prefixed with '0') and second from recipient (prefixed with '1'). The following while loop will continue until an index for two genes prefixed with '0' and '1' is selected. At each iteration the chances the while loop must continue reduce exponentially since the index is chosen randomly form the same set of options.